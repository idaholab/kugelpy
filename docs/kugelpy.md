Copyright 2024, Battelle Energy Alliance, LLC, ALL RIGHTS RESERVED
# `kugelpy`

`kugelpy` is a python module that performs time-dependent high-temperature gas-cooled pebble bed reactor (PBR) analysis. 
This process can simulate the run-in procedure for a PBR along with equilibrium analysis. 
The following notebook will describe how to create a `kugelpy` input and how to load previous `kugelpy` simulations to restart or examine the state of the reactor.

The examples are run using a pbs script, see an example [here](./kugelpy/kugelpy.pbs), feel free to use this script and update it for your work. 
Typically the script should be run on a high performance computing cluster (HPC), as the simulations are computationally expensive (reasonable results have been shown between 30-40 nodes with 48 processors each).
The python script has been annotated to provide clarity for the input script.

## Examples

[Multi-Pass Jump-In Simulation](./jump_in.md)

[Multi-Pass Run-In Simulation](./run_in.md)

[OTTO Jump-In Simulation](./otto.md)

[Multiphysics Run-In Simulation](./mp_run_in.md)

[Loading a Previous Simulation](./loader.md)

### Variables

#### Questions

If you have questions or need help with the `kugelpy` module, please direct questions to Ryan Stewart (ryan.stewart@inl.gov).