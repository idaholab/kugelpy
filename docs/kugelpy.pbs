#!/bin/bash
#PBS -l select=24:ncpus=48:mpiprocs=1
#PBS -N ri_up_geo
#PBS -l walltime=120:00:00
#PBS -ko
#PBS -j oe
#PBS -P neams
#PBS -l software=Serpent
#PBS -m abe
#
cat $PBS_NODEFILE
#
source /etc/profile.d/modules.sh
module load use.exp_ctl
module load serpent/2.1.32-intel-19.0
module load python
conda activate pyrates
#
cd $PBS_O_WORKDIR
#
JOB_NUM=${PBS_JOBID%\.*}
if [ $PBS_O_WORKDIR != $HOME ]
then
ln -s $HOME/$PBS_JOBNAME.o$JOB_NUM $PBS_JOBNAME.o$JOB_NUM
fi
#
export TMPDIR=/tmp
#
#
#
python3 python_script.py -omp 48
#
#
if [ $PBS_O_WORKDIR != $HOME ]
then
rm $PBS_O_WORKDIR/$PBS_JOBNAME.o$JOB_NUM
mv $HOME/$PBS_JOBNAME.o$JOB_NUM $PBS_O_WORKDIR/$PBS_JOBNAME.o$JOB_NUM
fi