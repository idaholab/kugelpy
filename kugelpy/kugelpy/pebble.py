'''
@authors: balep, stewryan, ethan-fowler

Copyright 2024, Battelle Energy Alliance, LLC, ALL RIGHTS RESERVED
'''

import numpy as np
import math

class Pebble(object):
    """
    Class for holding information for an individual pebble.
    
    Parameters
    ----------
    x: float
        current x position in the pebble bed
    y: float
        current y position in the pebble bed
    z: float
        current z position in the pebble bed
    r: float
        current radial position in the pebble bed
    radius: float
        Outer radius of the pebble surrounding the graphite matrix [cm]
    volume: int
        Integer representation the volume region the pebble is in 
    channel: int
        Integer representation the channel the pebble is in     
    
    Attributes
    ----------
    pebble_type: str
        Designation if this is a graphite or fuel pebble
    inner_radius: float
        Inner radius of the pebble surrounding the graphite matrix [cm]
    material: dict
        Dictionary of the material in the pebble. Takes the form:
        {'material_name': {'ZAID1': float, 'ZAID2': float, ...}}
    geometry: dict
        Dictionary of the geometry in the pebble. 
        Each entry describes a part of the geometry, where a given radius and number of instances are used to calculate the volume
        Takes the form:
        {'geometry_name': {'radius': float, 'number_of_instances': int, 'volume': float}}
    x: float
        current x position in the pebble bed
    y: float
        current y position in the pebble bed
    z: float
        current z position in the pebble bed
    r: float
        current radial position in the pebble bed
    radius: float
        Outer radius of the pebble surrounding the graphite matrix [cm]
    volume_num: int
        Integer representation the volume region the pebble is in 
    channel_num: int
        Integer representation the channel the pebble is in 
    mesh_location: str
        Combination of channel, and volume for where the pebble is
    xs_library: str
        Cross-section library for Serpent to use 
    scattering_library: str
        Name of the scattering library for Serpent to use
    temperature: float
        Temperature of the pebble [K]
    shuffled: Bool
        Indicator if this is the first pass or Nth pass through the core
    previous_universe: str
        Concatinated string containing the previous universes the pebble was exposed to
        If the pebble is homogenized after each pass, we default the previous universes to c0v0
    burnup: float
        Burnup accrued during pebble movement [MWd/kg]
    num_passes: int
        number of passes a pebble has gone through the core
    pebble_number: int
        integer related to the pebble number in the core
    universe: str
        string containing two integers, channel and volume
    """

    def __init__(self, x, y, z, r, radius, channel, volume, temperature=900.0, pass_limit=6, pebble_number=0):
        self._pebble_type = 'graphite'
        self._inner_radius = 2.5
        self._radius = radius
        self._material = {'matrix':   {'6000': 9.22570E-2, '5010': 2.28336E-9, '5011': 9.24876E-9},
                         'pebshell': {'6000': 6.61814E-2, '5010': 2.23273E-8,  '5011': 9.04368E-7}}
#        self._material = {'matrix':   {'6000': 9.22570E-2, },
#                         'pebshell': {'6000': 6.61814E-2, }}        
        self._x = x
        self._y = y
        self._z = z
        self._r = r
        self._volume_num = volume
        self._channel_num = channel
        self._mesh_location = f'c{channel}v{volume}'
        self._temperature = temperature
        self._shuffled = False
        self._previous_universe = ''
        self._burnup = 0.
        self._burnup_j_cm3 = 0.
#        self._burnup_group = 0
        self._num_passes = 0
        self._pass_limit = pass_limit
        self._pebble_number = pebble_number
        self._xs_dict = {1500: {'graphite': 'grph1500', 'xs_set': '15c'},
                        1200: {'graphite': 'grph1200', 'xs_set': '12c'},
                         900: {'graphite':  'grph900', 'xs_set': '09c'},
                         600: {'graphite':  'grph600', 'xs_set': '06c'},
                         300: {'graphite':  'grph300', 'xs_set': '03c'},}
        self._previous_universe = 'g0_' + self._mesh_location
        self.set_universe()
        self._xs_library, self._scattering_library = self.set_xs_set(self._temperature)

        self._geometry = {'matrix':    {'radius': self._inner_radius,    'number_of_instance': 1, 'volume': 0.0},
                         'pebshell':  {'radius': radius, 'number_of_instance': 1, 'volume': 0.0}}
        self.calculate_volumes()

    def calculate_volumes(self):
        """
        Calculate the volume for each constituent part of the pebble
        """
        previous_volume = 0.0
        for data in self._geometry.values():
            volume = math.pi * 4 / 3 * math.pow(data['radius'], 3) 
            total_region_volume = volume * data['number_of_instance']
            data['volume'] = total_region_volume - previous_volume
            previous_volume = total_region_volume 

    def set_xs_set(self, temperature):
        """
        Grab the correct cross-section set and scattering library based on the temperature
        """
        dict_array = np.array([x for x in self._xs_dict.keys()])
        temp = dict_array[dict_array<temperature].max() if temperature not in self._xs_dict.keys() else temperature
        return self._xs_dict[temp]['xs_set'], self._xs_dict[temp]['graphite']
        
    def update_position(self, x, y, z, r, channel, volume, shuffled=False):
        """
        Update the position of the pebbles after a shift or refuel.
        """
        self._shuffled = shuffled
        self._x = x
        self._y = y
        self._z = z
        self._r = r
        self._channel_num = channel
        self._volume_num = volume
        self._mesh_location = f'c{channel}v{volume}'
        self.set_universe()

    def set_previous_universe(self, prev_universe):
        """
        Update the previous universe after a shift or refuel
        """
        self._previous_universe = prev_universe

    def increase_pass(self):
        """
        When pebble is refueled, increse the number of passes it has gone through.
        """
        self._num_passes += 1
    
    def set_universe(self):
        """
        Update the current universe of the pebbble, if this is not the first pass (i.e. it has been shuffled, add the previous universe to the current pebble location
        """
        if self._shuffled:
            self._universe = self._previous_universe + '_' + self._mesh_location
        else:
            split_univ = self._previous_universe.split('_')[1:]
            univ = '' if len(split_univ) == 1 else '_'.join([x for x in split_univ[:-1]]) + '_'
            self._universe = f'g_' + univ + self._mesh_location 
            #self._universe = f'g_' + f'p{self._num_passes}_' + f'bu{self._burnup_group}_ + univ + self._mesh_location
    
    def update_pebble_temperature(self,pebble_temp,fuel_temp=None):
        """
        Update the pebble temperature and corresponding cross-section/scattering libraries.
        """
        self._temperature = pebble_temp
        self._xs_library, self._scattering_library = self.set_xs_set(pebble_temp)

class FuelPebble(Pebble):
    """
    Class for holding information for an individual pebble.
    
    Parameters
    ----------
    x: float
        current x position in the pebble bed
    y: float
        current y position in the pebble bed
    z: float
        current z position in the pebble bed
    r: float
        current radial position in the pebble bed
    radius: float
        Outer radius of the pebble surrounding the graphite matrix [cm]
    volume: int
        Integer represention the volume region the pebble is in 
    channel: int
        Integer represention the channel the pebble is in 
    fuel_material: dict
        Dictionary containing the fuel material, will replace 'fuel' in self._material
    
    
    Attributes
    ----------
    kernels_per_pebble: int
        Number of TRISO kernels per fuel pebble
    triso_volume: float
        Volume of the fuel kernels in the pebble (required for correct burnup)
    power_days: float
        Number of megawatt days the pebbles has been in the core (days * power) [MWd]
    power_density: float
        Current power density for the pebble (this is the power density for all pebbles of the same fuel type, same pass number, and in the same core location) [MW]
    geometry: dict

    """
    
    def __init__(self, x, y, z, r, radius, channel, volume, fuel_material=None, homogenization_group=0, temperature=900.0,  fuel_temperature=900.0, pass_limit=6, pebble_number=0, kernel_data=None):
        self._homogenization_group = homogenization_group
        super().__init__(x, y, z, r, radius, channel, volume, temperature=temperature, pass_limit=pass_limit, pebble_number=pebble_number)
        self._pebble_type = 'fuel'
        self._kernel_data = kernel_data if kernel_data else {'fuel':0.02125, 'buffer':0.03125, 'inner_pyc':0.03525, 'sic':0.03875, 'outer_pyc':0.04275, 'kernels_per_pebble': 18775}
        self._kernels_per_pebble = self._kernel_data.pop('kernels_per_pebble')
        self._material = {'fuel':      {'92235': 3.70100E-3, '92238': 1.99216E-2,  '8016': 3.37331E-1, '6000': 9.26006E-3, '5010': 1.87091E-8, '5011': 7.57813E-8},
                         'buffer':    { '6000': 5.26466E-2,  '5010': 1.35512E-8,  '5011': 5.48894E-8},
                         'inner_pyc': { '6000': 9.52653E-2,  '5010': 2.45213E-8,  '5011': 9.93236E-8},
                         'sic':       {'14028': 4.43271E-2, '14029': 2.25082E-3, '14030': 1.48376E-3, '6000': 4.80616E-2,  '5010': 1.23711E-8,  '5011': 5.0109E-8},
                         'outer_pyc': { '6000': 9.52653E-2,  '5010': 2.45213E-8,  '5011': 9.93236E-8},
                         'matrix':    { '6000': 8.67416E-2,  '5010': 2.23273E-8,  '5011': 9.04368E-8},
                         'pebshell':  { '6000': 8.67416E-2,  '5010': 2.23273E-8,  '5011': 9.04368E-8}}

        #self._material = {'fuel':      {'92235': 3.70100E-3, '92238': 1.99216E-2,  '8016': 3.37331E-1, '6000': 9.26006E-3, },
        #                 'buffer':    { '6000': 5.26466E-2, },
        #                 'inner_pyc': { '6000': 9.52653E-2,  },
        #                 'sic':       {'14028': 4.43271E-2, '14029': 2.25082E-3, '14030': 1.48376E-3, '6000': 4.80616E-2, },
        #                 'outer_pyc': { '6000': 9.52653E-2, },
        #                 'matrix':    { '6000': 8.67416E-2, },
        #                 'pebshell':  { '6000': 8.67416E-2, }}                         
        self._fuel_temperature = fuel_temperature
        self._previous_universe = f'f{self._homogenization_group}_' + self._mesh_location

        self.set_universe()
        self._previous_universe = self._universe
        self._triso_volume = self._kernels_per_pebble * 0.0000402
        self._power_days = 0.
        self._power_density = 0.
        self._days_in_core = 0
        if fuel_material:
            self._material['fuel'] = fuel_material
        self._xs_fuel_library, self._fuel_scattering_library = self.set_xs_set(self._fuel_temperature)

        self._geometry = {'fuel':      {'radius': self._kernel_data['fuel'],      'number_of_instance': self._kernels_per_pebble, 'volume': 0},
                         'buffer':    {'radius': self._kernel_data['buffer'],    'number_of_instance': self._kernels_per_pebble, 'volume': 0},
                         'inner_pyc': {'radius': self._kernel_data['inner_pyc'], 'number_of_instance': self._kernels_per_pebble, 'volume': 0},
                         'sic':       {'radius': self._kernel_data['sic'],       'number_of_instance': self._kernels_per_pebble, 'volume': 0},
                         'outer_pyc': {'radius': self._kernel_data['outer_pyc'], 'number_of_instance': self._kernels_per_pebble, 'volume': 0},
                         'matrix':    {'radius': self._inner_radius, 'number_of_instance': 1},
                         'pebshell':  {'radius': radius, 'number_of_instance': 1}}
        self.calculate_volumes()
        

    def update_fuel_material(self, material):
        """
        Update the fuel material
        """
        self._material['fuel'] = material

    def update_burnup(self, power, days):
        """
        Add up the total power days the pebble has been in the core.
        Divide by the kg of HM originally in the pebble (currently assumed to be 7 g or 0.007 kg)
        To obtain the burnup in J/cm^3 
        """
        self._days_in_core += days
        self._power_density = power / (1E6) # Convert from W to MW, power of an individual pebble
        self._power_days += (self._power_density * days)
        self._burnup =  self._power_days / 0.007 # MWd/kg
        self._burnup_j_cm3 = self._power_days / self._triso_volume * 1E6 * 86400# for J/cm^3 we divide the power days (Watt-day or J/s * days) by the kernel volume (cm^3) and convert to MJ from J (1E6) and from days to seconds (86,400 s/day)
        
    def set_universe(self):
        """
        Update the current universe of the pebbble, if this is not the first pass (i.e. it has been shuffled, add the previous universe to the current pebble location        
        """
        if self._shuffled:
            self._universe = self._previous_universe + '_' + self._mesh_location
        else:
            split_univ = self._previous_universe.split('_')[1:]
            univ = '' if len(split_univ) == 1 else '_'.join([x for x in split_univ[:-1]]) + '_'
            self._universe = f'f{self._homogenization_group}_' + univ + self._mesh_location   
            
    def update_pebble_temperature(self,pebble_temp,fuel_temp=None):
        """
        Update the temperatures (and corresponding cross-sections) for both the pebble and fuel materials.
        """
        self._temperature = pebble_temp
        self._fuel_temperature = fuel_temp
        self._xs_fuel_library, self._fuel_scattering_library = self.set_xs_set(fuel_temp)
        self._xs_library, self._scattering_library = self.set_xs_set(pebble_temp)
