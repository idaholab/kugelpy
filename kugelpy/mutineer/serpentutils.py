'''
Created on mar 30, 2020
@author: balep, stewryan, ethan-fowler

Copyright 2024, Battelle Energy Alliance, LLC, ALL RIGHTS RESERVED
'''

import os
from re import sub
import numpy as np

def bu_reader(path_to_bu):
    '''
    Given a file directory, grab each `bumat` file and return a dictionary based on the burnup step and pebble compositions
    '''
    files = [x[2] for x in os.walk(path_to_bu)]
    pebble_dict = {}
    for file_name in files[0]:
        if 'bumat' in file_name:
            #time is denotes by `bumat#`, where the # gives the time step
            time = file_name.split('bumat')
            time = int(time[-1])
            a_dens = nuclide_loop(path_to_bu, file_name)
            pebble_dict[time] = a_dens
    return pebble_dict

def nuclide_loop(path_to_bu, file_name):
    '''
    Extract the atom densities for each pebble   
    '''
    f = open(f'{path_to_bu}/{file_name}', "r")
    nuclide_dict = {}
    correct_lines = False
    for line in f:
        if 'fuel' in line or 'uco' in line:
            correct_lines = True
            mat, a_density, volume = line.split('  ')[1], float(line.split(' ')[-3]), float(line.split(' ')[-1])
            nuclide_dict[mat] = {}
            nuclide_dict[mat]['volume'] = volume
        elif correct_lines:
            nuclide, a_den = read_nuclide_atom_density(line)
            nuclide_dict[mat][nuclide] = a_den
            
    f.close()  
    return nuclide_dict

def read_nuclide_atom_density(line):
    '''
    Read the atomic densities in the `bumat` file
    '''
    split_line = line.split('  ')
    nuclide = str(int(split_line[-2].split('.')[0]))
    a_den = float(split_line[-1])
    return nuclide, a_den


def serpent_rd(file_path, xs_name_list, material_ids):
    '''
    Extract Data from Serpent 2.0 output file.
    '''
    # Do not process all the line when looking for headers.
    line_cutoff = 42
    # initialize the line byte counter.
    line_byte = 0
    # initialize dictionaries.
    data_info = {}
    for xs_name in xs_name_list:
        data_info[xs_name] = []
    # First Reading to determine data position.
    with open(file_path, 'r') as sf:
        for line in sf:
            # cutoff very long lines.
            line_len = len(line)
            if line_len > line_cutoff:
                line = line[:line_cutoff]
            # split the line to check the first word.
            line_split = line.split()
            # If the split worked.
            if line_split:
                for xs_name in xs_name_list:
                    if line_split[0] == xs_name.strip():
                        for word in line_split:
                            if ')' in word:
                                numf = int(''.join([x for x in word if x not in ')]']))
                        data_info[xs_name].append([line_byte, numf])
                        break
            # Update line byte position counter.
            line_byte += line_len
    # Initialize dictionaries.
    data = {}
    for xs_name in xs_name_list:
        data[xs_name] = []
    # Second reading to grab the data.
    with open(file_path, 'r') as sf:
        for xid in material_ids:
            for xs_name in xs_name_list:
                if len(data_info[xs_name]) == 0:
                    raise RuntimeError(xs_name + ' not avaliable...')
                sf.seek(data_info[xs_name][xid][0])
                dummy, values = sf.readline().split('=', 1)
                values = sub(r"\[|\]|;", "", values).strip()
                try:
                    float(values[0])
                    values = np.fromstring(values, dtype=float, sep=' ')
                except ValueError:
                    ...
                data[xs_name].append(values)
    # Elaborate data.
    for xs_name in xs_name_list:
        for id_x in range(len(material_ids)):
            if xs_name in ['PRECURSOR_GROUPS', 'MACRO_NG']:
                data[xs_name][id_x] = int(data[xs_name][id_x][0])
            elif xs_name in ['GC_UNIVERSE_NAME']:
                data[xs_name][id_x] = sub("'", "", data[xs_name][id_x]).strip()
            else:
                data[xs_name][id_x] = data[xs_name][id_x][0::2]

    # Return  the materials cross sections.
    return data
