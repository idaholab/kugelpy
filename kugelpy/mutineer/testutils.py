'''Utility module for testing modules.

Created on 2023-05-11 16:13:01

@authors: balep, stewryan, ethan-fowler

Copyright 2024, Battelle Energy Alliance, LLC, ALL RIGHTS RESERVED
'''
#===============================================================================
# Import
#===============================================================================
from _ctypes import ArgumentError
from os import path, mkdir,getenv
from errno import EEXIST

#===============================================================================
# Definitions
#===============================================================================
def gen_tmp_folder(main_tests_dir):
    if not path.isdir(main_tests_dir):
        try:
            mkdir(main_tests_dir)
        except OSError as exc:
            if exc.errno != EEXIST:
                raise
            pass

def find_pbed_input(file, file_path):
    '''
    Attempts to locate the start, end, and format of Serpent pbed position input in file.

    file: (str) name of file with pbed data 
    file_path: (str) path to file
    '''

    found_start = False # keeps track of whether or not start index had been found
    start_line = None # stores fist pbed geometry input line index
    end_line = None # stores last pbed geometry input line index
    format = None # indicates format of pbed geometry input line

    input = open(path.join(file_path, file)).readlines()
    
    for i in range(len(input)):
        temp = input[i].split()
        if len(temp)==6:
            try:
                float(temp[5])
            except:
                if not found_start:
                    found_start = True
                    format = 1
                    start_line = i
        elif len(temp)==5:
            try:
                float(temp[4])
            except:
                if not found_start:
                    found_start = True
                    start_line = i
                    format = 0
        else:
            if found_start:
                end_line = i
    
    return start_line, end_line, format

def compare_pbeds(solution_file, solution_path, test_file, test_path, solution_start_line=0, solution_end_line=None, format=0):
    '''
    Reads in the position (x, y, z) and radius of pebbles for comparison with expectation.
    In some edge cases, variable rounding due to numpy version can result in pebbles being placed in different 
    universes, thus universe names are not accounted for.

    solution_file: (str) file name of pre-generated solution output used for comparison
    solution_path: (str) path to solution output file
    test_file: (str) file name of pre-generated test output used for comparison
    test_path: (str) path to test output file
    solution_start_line: (str) index of line in solution file where pebble positions begin (0 by default)
    solution_end_line: (str) index of line in solution file where pebble positions end (None by default)
    format: (int) 1 if lines are numbered, 0 otherwise
    '''

    # read in solution and test files
    solution = open(path.join(solution_path, solution_file)).readlines()[solution_start_line:solution_end_line]
    test = open(path.join(test_path, test_file)).readlines()
    
    # remove new line characters
    while '\n' in solution:
        solution.remove('\n')
    while '\n' in test:
        test.remove('\n') 

    if len(solution) != len(test):
        # if files contain different number of pebbles, write failed
        print("compare_pbeds() Error: Different number of pebbles")
        print("Solution Length: ", len(solution))
        print("Test Length: ", len(test))
        return False
    else:
        # format solution and test lists
        for i in range(len(solution)):
            temp = solution[i].split()
            solution[i] = f'{temp[0+format]} {temp[1+format]} {temp[2+format]} {temp[3+format]}'
            temp = test[i].split()
            test[i] = f'{temp[0+format]} {temp[1+format]} {temp[2+format]} {temp[3+format]}'
        
        for line in test:
            # check to make sure every pebble in test is also in solution
            if line in solution:
                # remove line from solution to reduce number of comparison operations
                solution.remove(line)
            else:
                # pebble in line not contained in solution, write failed
                print("Incorrect line: ", line)
                print("Remaining length: ", len(solution))
                return False
        
        return True

#===============================================================================
# Main
#===============================================================================
if __name__ == '__main__':
    pass

