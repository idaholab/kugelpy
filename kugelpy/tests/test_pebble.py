'''
@author: balep, stewryan, ethan-fowler

Copyright 2024, Battelle Energy Alliance, LLC, ALL RIGHTS RESERVED
'''

import kugelpy.kugelpy.kugelpy.pebble as pebble
from kugelpy.kugelpy.mutineer.testutils import gen_tmp_folder
from os import path

main_dir = path.dirname(path.realpath(__file__))
ifiles_dir = path.join(main_dir, 'testfiles')
main_tests_dir = path.join(main_dir,  '_tmp/')
gen_tmp_folder(main_tests_dir)

def test_init(regtest):
    peb = pebble.Pebble(111,111,111,111,111,4,5,temperature=600,pass_limit=5,pebble_number=12345)
    print(vars(peb), file=regtest)

def test_update_position(regtest):
    peb = pebble.Pebble(1,1,1,1,1,1,2,temperature=600,pass_limit=5,pebble_number=12345)
    peb.update_position(111,111,111,111,4,5,shuffled=True)
    print(vars(peb), file=regtest)

def test_increase_pass(regtest):
    peb = pebble.Pebble(1,1,1,1,1,1,2,temperature=600,pass_limit=5,pebble_number=12345)
    peb.increase_pass()
    print(vars(peb), file=regtest)

def test_update_pebble_temperature(regtest):
    peb = pebble.Pebble(1,1,1,1,1,1,2,temperature=600,pass_limit=5,pebble_number=12345)
    peb.update_pebble_temperature(547,fuel_temp=900)
    print(vars(peb), file=regtest)

def test_FuelPebble_init(regtest):
    peb = pebble.FuelPebble(111,111,111,111,111,4,5,temperature=600,fuel_temperature=576, pass_limit=5,pebble_number=12345)
    print(vars(peb), file=regtest)

def test_update_fuel_material(regtest):
    peb = pebble.FuelPebble(111,111,111,111,111,4,5,temperature=600,fuel_temperature=576, pass_limit=5,pebble_number=12345)
    peb.update_fuel_material({'fractions': {'1001': 1.0}})
    print(vars(peb), file=regtest)

def test_update_burnup(regtest):
    peb = pebble.FuelPebble(111,111,111,111,111,4,5,temperature=600,fuel_temperature=576, pass_limit=5,pebble_number=12345)
    peb.update_burnup(20, 5)
    print(vars(peb), file=regtest)

def test_update_pebble_temperature(regtest):
    peb = pebble.FuelPebble(111,111,111,111,111,4,5,temperature=600,fuel_temperature=576, pass_limit=5,pebble_number=12345)
    peb.update_pebble_temperature(547,fuel_temp=900)
    print(vars(peb), file=regtest)

def test_calculate_volumes(regtest):
    peb = pebble.Pebble(111,111,111,111,3,4,5)
    peb.calculate_volumes()
    print(peb._geometry, file=regtest)
    
    fpeb = pebble.FuelPebble(111,111,111,111,3,4,5)
    fpeb.calculate_volumes()
    print(fpeb._geometry, file=regtest)    
