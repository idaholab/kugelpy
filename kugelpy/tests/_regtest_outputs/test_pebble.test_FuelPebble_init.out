{'_homogenization_group': 0, '_pebble_type': 'fuel', '_inner_radius': 2.5, '_radius': 111, '_material': {'fuel': {'92235': 0.003701, '92238': 0.0199216, '8016': 0.337331, '6000': 0.00926006, '5010': 1.87091e-08, '5011': 7.57813e-08}, 'buffer': {'6000': 0.0526466, '5010': 1.35512e-08, '5011': 5.48894e-08}, 'inner_pyc': {'6000': 0.0952653, '5010': 2.45213e-08, '5011': 9.93236e-08}, 'sic': {'14028': 0.0443271, '14029': 0.00225082, '14030': 0.00148376, '6000': 0.0480616, '5010': 1.23711e-08, '5011': 5.0109e-08}, 'outer_pyc': {'6000': 0.0952653, '5010': 2.45213e-08, '5011': 9.93236e-08}, 'matrix': {'6000': 0.0867416, '5010': 2.23273e-08, '5011': 9.04368e-08}, 'pebshell': {'6000': 0.0867416, '5010': 2.23273e-08, '5011': 9.04368e-08}}, '_x': 111, '_y': 111, '_z': 111, '_r': 111, '_volume_num': 5, '_channel_num': 4, '_mesh_location': 'c4v5', '_temperature': 600, '_shuffled': False, '_previous_universe': 'f0_c4v5', '_burnup': 0.0, '_burnup_j_cm3': 0.0, '_num_passes': 0, '_pass_limit': 5, '_pebble_number': 12345, '_xs_dict': {1500: {'graphite': 'grph1500', 'xs_set': '15c'}, 1200: {'graphite': 'grph1200', 'xs_set': '12c'}, 900: {'graphite': 'grph900', 'xs_set': '09c'}, 600: {'graphite': 'grph600', 'xs_set': '06c'}, 300: {'graphite': 'grph300', 'xs_set': '03c'}}, '_universe': 'f0_c4v5', '_xs_library': '06c', '_scattering_library': 'grph600', '_geometry': {'fuel': {'radius': 0.02125, 'number_of_instance': 18775, 'volume': 0.7546496207696666}, 'buffer': {'radius': 0.03125, 'number_of_instance': 18775, 'volume': 1.6453911536097428}, 'inner_pyc': {'radius': 0.03525, 'number_of_instance': 18775, 'volume': 1.0446157118140604}, 'sic': {'radius': 0.03875, 'number_of_instance': 18775, 'volume': 1.1313188552168967}, 'outer_pyc': {'radius': 0.04275, 'number_of_instance': 18775, 'volume': 1.5683883222058599}, 'matrix': {'radius': 2.5, 'number_of_instance': 1, 'volume': 59.30548328617113}, 'pebshell': {'radius': 111, 'number_of_instance': 1, 'volume': 5728653.886715267}}, '_kernel_data': {'fuel': 0.02125, 'buffer': 0.03125, 'inner_pyc': 0.03525, 'sic': 0.03875, 'outer_pyc': 0.04275}, '_kernels_per_pebble': 18775, '_fuel_temperature': 576, '_triso_volume': 0.7547550000000001, '_power_days': 0.0, '_power_density': 0.0, '_days_in_core': 0, '_xs_fuel_library': '03c', '_fuel_scattering_library': 'grph300'}
